### PR TITLE
DAOS-15251 cq: add codespell githook

### PR DESCRIPTION
**Pull request by daltonbohning**
Add a codespell githook.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,3 +3,9 @@ supported_extensions = ["py"]
 skip = [".git/", "src/rdb/raft", "build", "install", "venv", "src/control/vendor/", "_topdir"]
 line_length = 99
 skip_gitignore = true
+
+[tool.codespell]
+skip = './src/control/vendor,./src/control/go.sum,./.git,./src/rdb/raft,./build,./install,./venv,./src/control/vendor/'
+quiet-level = 3
+ignore-words = 'ci/codespell.ignores'
+builtin = 'clear,rare,informal,names,en-GB_to_en-US'
```
Comment (daltonbohning) [inline pyproject.toml:11]: This lets us run codespell almost similar to the githook and the GHA. But unfortunately codespell doesn't seem to have any git option or way to skip files in .gitignore

File: utils/githooks/pre-commit.d/40-codespell.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    exit 0
+fi
+
+if [ "$TARGET" = "HEAD" ]; then
+        echo "  Checking against HEAD"
+        git diff HEAD --name-only | xargs codespell
+else
+        echo "  Checking against branch ${TARGET}"
+        git diff "$TARGET"... --name-only | xargs codespell
+fi
```
Comment (daltonbohning) [inline utils/githooks/pre-commit.d/40-codespell.sh:28]: I do wonder if it's faster to use xargs or something like this?
```
codespell $(git diff "$TARGET"... --name-only)
```
codespell itself doesn't support piping files

**Review by jolivier23 — APPROVED** [review]

**Review by phender — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (ashleypittman) [reply pyproject.toml:11]: Shame.  I had lots of failures when I updated to get this githook, many of them were from 'docs/doxygen` which is not included in the source but is the default build location for the documentation.

Comment (daltonbohning) [reply pyproject.toml:11]: Do we need to add docs/doxygen to the skip here?

Comment (ashleypittman) [reply pyproject.toml:11]: I was able to just delete the directory and carry on.  Like you say best would be to support a .gitignore file.

An option might be to use `git ls-files` or similar which would mean we only needed to skip committed files which should be ignored, not non-committed files as well.

Comment (daltonbohning) [reply pyproject.toml:11]: I thought `git diff` already ignored non-committed files?

Comment (ashleypittman) [reply pyproject.toml:11]: Oh yeah, I'm not sure why it was even testing that dir then.  Let me check again.

**Pull request merged**